docs/sync-overview.md

```diff
@@ -0,0 +1,57 @@
+
+### Sync manager
+
+[`SyncManager`](https://github.com/mozilla/application-services/blob/main/components/sync_manager/src/manager.rs) is responsible for performing the high-level parts of the sync process:
+  - The consumer code calls it's `sync()` function to start the sync, passing
+    in a `SyncParams` object in, which describes what should be synced.
+  - `SyncManager` checks if we are currently in a backoff period and should wait before contacting the server again.
+  - `SyncManager` creates invididual `SyncEngine` instances for each component that should be synced.  We currently use 2 different methods for this:
+    - The older method is for the `SyncManager` to store a weakref to the `Store` for each component and use that to create a `SyncEngine` (tabs and places)
+    - The newer method is for the components to provide a function to create the `SyncEngine`, hiding the details of how that engine is created (autofill/logins)
+  - For components that use local encryption, `SyncManager` passes the local encryption key to their `SyncEngine`
+  - Finally, calls `sync_multiple()` function from the `sync15` crate, sending it the `SyncEngine` instances.  `sync_multiple()` then calls the `sync()` function for each individual `SyncEngine`
+
+### Sync engines
+  - [`SyncEngine`](https://github.com/mozilla/application-services/blob/main/components/support/sync15-traits/src/engine.rs) is defined in the `support/sync15-traits` crate and defines the interface for syncing a component.
+  - A new `SyncEngine` instance is created for each sync
+  - `SyncEngine.apply_incoming()` does the main work.  It is responsible for processing incoming records from the server in order to update the local records and calculating which local records should be synced back.
+
+## The `apply_incoming` pattern
+
+`SyncEngine` instances are free to implement `apply_incoming()` any way they want, but the most components follow a general pattern.
+
+### Database Tables
+
+   - The local table stores records for the local application
+   - The mirror table stores the last known record from the server
+   - The staging temporary table stores the incoming records that we're currently processing
+   - Each table contains a `guid` as its primary key.  A record will share the same `guid` for the local/mirror/staging table.
+
+### `apply_incoming` stages
+  - **stage incoming**: write out each incoming server record to the staging table
+  - **fetch states**: take the rows from all 3 tables and combine them into a single struct containing `Option`s for the local/mirror/staging records.
+  - **iterate states**: loop through each state, decide how to do change the local records, then execute that plan.
+    - **reconcile/plan**: For each state we create an action plan for it.  The action plan is a low-level description of what to change (add this record, delete this one, modify this field, etc).  Here are some common situations:
```

Comment:
I'm super curious about this, I imagine each component has its own magic that it does for figuring out how to sync in different cases...

Maybe for the future (or maybe this already exists somewhere!), It would be awesome to have short descriptions of how different components handle different states... It's a lotta work, not sure where it would fit and possibly just looking at the code is suffienct so we might be okay without it 🤷

Reply:
Yeah this would be very nice to have.  If you look through the engines code it's usually pretty readable, but a couple rustdoc strings could go a long way and would also be linkable from this doc.